### PR TITLE
2106 Legend.symbol() doesn't pass in index correctly

### LINKED ISSUE #2106: Legend.symbol() doesn't pass in index correctly
```
legend.symbolFactoryAccessor(function (d, i) {
    console.log(i);
  });
```
logs:
0
0
0

### PR DESCRIPTION
**Pull request by ztsai**
index was wrong due to nested selection

closes #2106

### CODE REVIEW

**Comment by bluong** [comment]
Please make a pull request to DefinitelyTyped and cross reference so we have the consistent diff between the typings file in both places.

**Comment by bluong** [comment]
@jtlan made a comment on the previous iteration of the pull request that can essentially let you dodge the need to check the exact d attribute, which is possibly better.

All we want to check is the index of the entry, so let's verify only the index

File: src/components/legend.ts [hunk 1]
```diff
@@ -277,9 +277,9 @@ export module Components {
         });
       });
 
-      entries.select("path").attr("d", (d: any, i: number) => this.symbol()(d, i)(layout.textHeight * 0.6))
+      entries.select("path").attr("d", (d: any, i: number, j: number) => this.symbol()(d, j)(layout.textHeight * 0.6))
```
Comment (bluong) [inline src/components/legend.ts]: See if you can remove the explicit typings from here.  Not a big deal though.

File: test/components/legendTests.ts [hunk 2]
```diff
@@ -355,4 +355,20 @@ describe("Legend", () => {
...
     svg.remove();
   });
+
+  it("symbol() passes index correctly", () => {
+    var domain = ["AA", "BB", "CC"];
+    color.domain(domain);
+
+    var expectedIndex = 0;
+    var symbolChecker = (d: any, index: number) => {
+      assert.strictEqual(index, expectedIndex, "message");
```
Comment (bluong) [inline test/components/legendTests.ts]: The message here should be something that's more meangingful.

**Comment by bluong** [comment]
@jtlan Routing the assignment to you since I'm biased on this PR.

File: test/components/legendTests.ts (hunk 2 above)
Comment (jtlan) [inline test/components/legendTests.ts]: When I originally typed `"message"`, I originally meant that as an example ... I should have made myself clearer.

**Comment by aicioara** [comment]
@jtlan Just curious, why did we alter `d3.d.ts` ? It seems odd

**Comment by crmorford** [comment]
Congrats on your first plottable PR @ztsai :)

**Comment by ztsai** [comment]
@aicioara the diff looks different because our d3.d.ts was not up to date, but the raw files would be the same

**Comment by aicioara** [comment]
I see, thank you.

Also, congrats for the first PR @ztsai !

**Pull request merged**